File: src/trendlens_frontend/src/lib/request.ts
import type {
  GeneralBalanceRequest,
  GeneralInstrumentsRequest,
  GeneralPostOrderRequest,
  Request,
} from "../../../declarations/trendlens_backend/trendlens_backend.did";
import type { ValueOf } from "./apiAddition";

export enum RequestType {
  Empty = "Empty",
  GetInstruments = "Get instruments",
  GetBalance = "Get balance",
  PostOrder = "Post order",
}

export enum RequestPickState {
  ApiRegistered,
  RequestPicked,
  RequestFilled,
}

// should be generic on setters and getters
// but i use it in specific place where
// i can't infer those generics
export interface IEnum {
  isEnum(): boolean;
  getVariants(): string[];
  getName(): string;
  get v(): string;
  set v(v: string);
}

export enum OrderSideType {
  Buy = "Buy",
  Sell = "Sell",
}

export class OrderSide implements IEnum {
  public variant: OrderSideType;

  constructor(t: OrderSideType) {
    this.variant = t;
  }

  get v() {
    return this.variant;
  }

  set v(v: OrderSideType) {
    this.variant = v;
  }

  public getVariants() {
    return Object.keys(OrderSideType);
  }

  public getName() {
    return "order side";
  }

  public isEnum() {
    return true;
  }
}

export enum TradeModeType {
  Cash = "Cash",
  SpotIsolated = "SpotIsolated",
  Isolated = "Isolated",
  Cross = "Cross",
}

export enum OrderTypeType {
  Market = "Market",
  Limit = "Limit",
  Ioc = "Ioc",
  Fok = "Fok",
  PostOnly = "PostOnly",
}

export enum PositionSideType {
  Short = "Short",
  Long = "Long",
}

export class OrderType implements IEnum {
  public variant: OrderTypeType;

  constructor(t: OrderTypeType) {
    this.variant = t;
  }

  get v() {
    return this.variant;
  }

  set v(v: OrderTypeType) {
    this.variant = v;
  }

  public getVariants() {
    return Object.keys(OrderTypeType);
  }

  public getName() {
    return "order type";
  }

  public isEnum() {
    return true;
  }
}

export class TradeMode implements IEnum {
  public variant: TradeModeType;

  constructor(t: TradeModeType) {
    this.variant = t;
  }

  get v() {
    return this.variant;
  }

  set v(v: TradeModeType) {
    this.variant = v;
  }

  public getVariants() {
    return Object.keys(TradeModeType);
  }

  public getName() {
    return "trade mode";
  }

  public isEnum() {
    return true;
  }
}

export class PositionSide implements IEnum {
  public variant: PositionSideType;

  constructor(t: PositionSideType) {
    this.variant = t;
  }

  get v() {
    return this.variant;
  }

  set v(v: PositionSideType) {
    this.variant = v;
  }

  public getVariants() {
    return Object.keys(PositionSideType);
  }

  public getName() {
    return "position side";
  }

  public isEnum() {
    return true;
  }
}

export class Instruments implements IEnum {
  public variant: InstrumentType;

  constructor(t: InstrumentType) {
    this.variant = t;
  }

  get v() {
    return this.variant;
  }

  set v(v: InstrumentType) {
    this.variant = v;
  }

  public getVariants() {
    return Object.keys(InstrumentType);
  }

  public getName() {
    return "instrument type";
  }

  public isEnum() {
    return true;
  }
}

export enum InstrumentType {
  Spot = "Spot",
  Features = "Features",
  Swap = "Swap",
  Margin = "Margin",
}

type BaseRequest = {
  type: RequestType;
};

export type InstrumentsRequest = {
  instrumentType: Instruments;
  instrumentId: OptionalField<string>;
} & BaseRequest;

export type BalanceRequests = {
  currencies: OptionalField<string>;
} & BaseRequest;

export interface IOptional<T> {
  value: T | null;
}

export class OptionalField<T> implements IOptional<T> {
  public value: T | null;

  constructor(v: T | null) {
    this.value = v;
  }
}

export type PostOrderRequest = {
  marginCurrency: OptionalField<string>;
  size: number;
  instrumentId: string;
  orderPrice: OptionalField<number>;
  side: OrderSide;
  tradeMode: TradeMode;
  orderType: OrderType;
  positionSide: OptionalField<PositionSide>;
} & BaseRequest;

export type ExchangeRequest =
  | InstrumentsRequest
  | BalanceRequests
  | PostOrderRequest;


export function isInstrumentsRequest(
  request: Request,
): request is { Instruments: GeneralInstrumentsRequest } {
  return (
    (request as { Instruments: GeneralInstrumentsRequest }).Instruments !==
    undefined
  );
}

export function isPostOrderRequest(
  request: Request,
): request is { PostOrder: GeneralPostOrderRequest } {
  return (
    (request as { PostOrder: GeneralPostOrderRequest }).PostOrder !== undefined
  );
}

export function isBalancesRequest(
  request: Request,
): request is { Balances: GeneralBalanceRequest } {
  return (
    (request as { Balances: GeneralBalanceRequest }).Balances !== undefined
  );
}

File: src/trendlens_backend/src/remote_exchanges/request.rs
use super::okx::api::InstrumentType;
use crate::pair::Pair;
use candid::CandidType;
use serde::{Deserialize, Serialize};
use std::{fmt::Display, str::FromStr};

#[derive(Debug, Clone, Deserialize, Serialize, CandidType, Copy)]
pub enum OrderSide {
    Buy,
    Sell,
}

impl Display for OrderSide {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        match self {
            OrderSide::Buy => write!(f, "buy"),
            OrderSide::Sell => write!(f, "sell"),
        }
    }
}

impl FromStr for OrderSide {
    type Err = String;
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "buy" => Ok(OrderSide::Buy),
            "sell" => Ok(OrderSide::Sell),
            _ => Err(format!("Unknown order side: {}", s)),
        }
    }
}

#[derive(Debug, Clone, Deserialize, Serialize, CandidType, Copy)]
pub enum PositionSide {
    Short,
    Long,
}

// no conditional bounding here, but with
// adding another exchange this should be hierarchical
#[derive(Debug, Clone, Deserialize, Serialize, CandidType, Copy)]
pub enum TradeMode {
    // Margin trading
    Cross,
    Isolated,

    // Non-margin trading
    Cash,
    SpotIsolated,
}

impl Display for TradeMode {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        match self {
            TradeMode::Cross => write!(f, "cross"),
            TradeMode::Isolated => write!(f, "isolated"),
            TradeMode::Cash => write!(f, "cash"),
            TradeMode::SpotIsolated => write!(f, "spot_isolated"),
        }
    }
}

impl FromStr for TradeMode {
    type Err = String;
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "cross" => Ok(TradeMode::Cross),
            "isolated" => Ok(TradeMode::Isolated),
            "cash" => Ok(TradeMode::Cash),
            "spot_isolated" => Ok(TradeMode::SpotIsolated),
            _ => Err(format!("Unknown trade mode: {}", s)),
        }
    }
}

#[derive(Debug, Clone, Deserialize, Serialize, CandidType, Copy)]
pub enum OrderType {
    Market,
    Limit,
    PostOnly,
    Fok,
    Ioc,
    // there are more
}


impl Display for OrderType {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        match self {
            OrderType::Market => write!(f, "market"),
            OrderType::Limit => write!(f, "limit"),
            OrderType::PostOnly => write!(f, "post_only"),
            OrderType::Fok => write!(f, "fok"),
            OrderType::Ioc => write!(f, "ioc"),
        }
    }
}

impl FromStr for OrderType {
    type Err = String;
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "market" => Ok(OrderType::Market),
            "limit" => Ok(OrderType::Limit),
            "post_only" => Ok(OrderType::PostOnly),
            "fok" => Ok(OrderType::Fok),
            "ioc" => Ok(OrderType::Ioc),
            _ => Err(format!("Unknown order type: {}", s)),
        }
    }
}

#[derive(Debug, Clone, Deserialize, Serialize, CandidType)]
pub struct GeneralPostOrderRequest {
    pub instrument_id: Pair,
    pub trade_mode: TradeMode,
    pub side: OrderSide,
    pub margin_currency: Option<String>,
    pub position_side: Option<PositionSide>,
    pub order_type: OrderType,
    pub size: f64,
    pub order_price: Option<f64>,
}

#[derive(Debug, Clone, Deserialize, Serialize, CandidType)]
pub struct GeneralInstrumentsRequest {
    pub instrument_type: InstrumentType,
    pub instrument_id: Option<Pair>,
}

#[derive(Debug, Clone, Deserialize, Serialize, CandidType)]
pub struct GeneralBalanceRequest {
    pub currency: Option<Vec<String>>,
}

#[derive(Debug, Clone, Deserialize, Serialize, CandidType)]
pub struct GeneralOrdersListRequest {
    pub instrument_type: InstrumentType,
    pub instrument_id: Pair,
    pub pending: bool,
}
